scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/constraints/activities/ActivityExpression.java

```diff
@@ -614,4 +616,76 @@ public String prettyPrint(final String prefix) {
   @Override
   public void extractResources(final Set<String> names) { }
 
+  //true if second is a subset or equal to first
+  public static boolean subsetOrEqual(SerializedValue first, SerializedValue second){
+    final var visitor = new SerializedValue.Visitor<Boolean>(){
+      @Override
+      public Boolean onNull() {
+        return true;
+      }
+
+      @Override
+      public Boolean onNumeric(final BigDecimal value) {
+        final var firstAsMap = first.asNumeric();
+        if(firstAsMap.isEmpty()){
+          return false;
+        }
+        return firstAsMap.get().equals(value);
+      }
+
+      @Override
+      public Boolean onBoolean(final boolean value) {
+        final var firstAsBoolean = first.asBoolean();
+        if(firstAsBoolean.isEmpty()){
+          return false;
+        }
+        return firstAsBoolean.get().equals(value);
+      }
+
+      @Override
+      public Boolean onString(final String value) {
+        final var firstAsString = first.asString();
+        if(firstAsString.isEmpty()){
+          return false;
+        }
+        return firstAsString.get().equals(value);
+      }
+
+      @Override
+      public Boolean onMap(final Map<String, SerializedValue> value) {
+        final var firstAsMap = first.asMap();
+        if(firstAsMap.isEmpty()){
+          return false;
+        }
+        for(final var elementInSecond: value.entrySet()){
+          final var elementInFirst = firstAsMap.get().get(elementInSecond.getKey());
+          if(elementInFirst != null){
```

Comment:
As you said, there might be elements in second that are not in first.

Reply:
If there is a `someKey: null` element, the subsequent `subsetOrEqual` call will fail as the matching will call a `asSomeType` method on it. So I guess it needs to be protected.

Reply:
I do agree that `containsKey` is a more direct way to communicate what is going on here. (The fact that `get` on `Map` returns null makes me sad every time...)

I'll leave it up to you what you think is more readable - here's what came out when I moved code around in my editor for a few minutes:
```java
     @Override
      public Boolean onMap(final Map<String, SerializedValue> value) {
        return arguments
            .asMap()
            .map(argumentsAsMap -> value
                .entrySet()
                .stream()
                .allMatch(patternElement -> {
                  if (!argumentsAsMap.containsKey(patternElement.getKey())) return false;
                  return subsetOrEqual(argumentsAsMap.get(patternElement.getKey()), patternElement.getValue());
                }))
            .orElse(false);
      }
```

Reply:
Or, if we wanted to use `&&` instead of `if`:

```java
      @Override
      public Boolean onMap(final Map<String, SerializedValue> value) {
        return arguments
            .asMap()
            .map(argumentsAsMap -> value
                .entrySet()
                .stream()
                .allMatch(
                    patternElement ->
                        argumentsAsMap.containsKey(patternElement.getKey()) 
                        && subsetOrEqual(argumentsAsMap.get(patternElement.getKey()), patternElement.getValue())))
            .orElse(false);
      }
```

Reply:
I have added this at the beginning of `subsetOrEqual`: 
```
   if(pattern == null){
      return true;
    }
    if(arguments == null){
      return pattern.isNull();
    }
```
and a test in `SerializedValueSubsetTest`.

It is tricky, here I say that `SerializedValue.NULL` is a subset of `null` but that could be the subject of a discussion.

Reply:
We had a discussion with @mattdailis and the conclusion is that neither arguments should be null, so I check them with `Objects.requireNonNull`.